projects/chaos/test/test_relay.c

```diff
@@ -0,0 +1,126 @@
+#include "relay.h"
+
+#include <inttypes.h>
+#include <stddef.h>
+
+#include "can.h"
+#include "can_ack.h"
+#include "can_msg_defs.h"
+#include "can_transmit.h"
+#include "chaos_events.h"
+#include "delay.h"
+#include "event_queue.h"
+#include "interrupt.h"
+#include "log.h"
+#include "relay_id.h"
+#include "soft_timer.h"
+#include "status.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#define NUM_CAN_RX_HANDLERS 2
+
+static CANStorage s_storage;
+static CANRxHandler s_rx_handlers[NUM_CAN_RX_HANDLERS];
+static CANAckRequests s_can_ack_requests;
+
+static StatusCode prv_rx_handler(const CANMessage *msg, void *context, CANAckStatus *ack_reply) {
+  CANAckStatus *status = context;
+  *ack_reply = *status;
+  LOG_DEBUG("Handled; responded with %u\n", *ack_reply);
+  return STATUS_CODE_OK;
+}
+
+void setup_test(void) {
+  event_queue_init();
+  interrupt_init();
+  soft_timer_init();
+
+  CANSettings settings = {
+    .device_id = CAN_DEVICE_CHAOS,
+    .bitrate = CAN_HW_BITRATE_125KBPS,
+    .rx_event = CHAOS_EVENT_CAN_RX,
+    .tx_event = CHAOS_EVENT_CAN_TX,
+    .fault_event = CHAOS_EVENT_CAN_FAULT,
+    .tx = { GPIO_PORT_A, 12 },
+    .rx = { GPIO_PORT_A, 11 },
+    .loopback = true,
+  };
+
+  can_init(&settings, &s_storage, s_rx_handlers, SIZEOF_ARRAY(s_rx_handlers));
+  can_ack_init(&s_can_ack_requests);
+
+  relay_init(true);
+}
+
+void teardown_test(void) {}
+
+typedef struct TestRelayEvents {
+  Event invoke_event;
+  Event response_event;
+} TestRelayEvents;
+
+typedef struct TestRelayParams {
+  RelayId id;
+  bool retry[2];
+} TestRelayParams;
+
+void test_relay_cycle(void) {
```

Comment:
I'm not sure what the scope of this test is, but it'd be nice to verify that we can

* Handle failure
* Handle success after retries
* Handle success
* Multiple concurrent relays - start one, start and finish another, finish the first one kind of thing. Just to make sure your FSMs don't somehow collide with each other or a transition occurs when it shouldn't.

Reply:
Done.